### PR TITLE
Diff converter v2

### PR DESCRIPTION
**Pull request by ArthurZucker**
# Introducing `diff_converter`, a linter to go from a modular file to a single model single file.

The `diff_converter` tool is here to replace our old `Copied from` statements, while keeping our core `transformers` philosophy:
- single model single file
- explicit code
- standardization of modeling code
- readable and educative code
- simple code
- least amount of modularity

This additionally unlocks the ability to very quickly see the differences between new architectures that get developed. While many architectures are similar, the "single model, single file" policy can obfuscate the changes. With this diff converter, we want to make the changes between architectures very explicit. 

## What it is:

It's a Python script that converts a `diff_my_model.py` file into `configuration_my_model.py` and `modeling_my_model.py` files using [AST](https://docs.python.org/3/library/ast.html). 

The `diff_my_model.py` file is a new way to write new models using inheritance (which had always been banned from `transformers`) to simplify model addition and improve readability. 

The **converter**, allows us to "unpack" the inheritance and obtain the traditional "single-file" `modeling` file. 
This diff is not to be imported from nor used directly, and isn't present in the inits of `transformers`. It will likely be renamed to `_diff_my_model.py` to clarify this; however, it can be used to debug the model implementation while contributing a new model.

## Supported syntax:

The `examples` folder holds examples of syntax that are usable, most notably:
- `super_kwargs` to replace with the parent class' kwargs
- `super().__function__()` calls, which will be replaced by the source code of the parent class, de-duplicating similar code
-  inheritance: base class dependencies are automatically fetched and written to the modeling file, including renamed decorators.

⚠️⚠️⚠️ This feature is in its infancy. It will be developed over the new few months in order to smoothen out the contribution of new models. ⚠️⚠️⚠️

If the conversion of you diff file does not go as planned / you encounter any issue with it, ping @ArthurZucker 🤗

### CODE REVIEW

File: src/transformers/models/gemma/modeling_gemma.py [hunk 1]
```diff
@@ -55,25 +50,30 @@
...
 
-# This makes `_prepare_4d_causal_attention_mask` a leaf function in the FX graph.
-# It means that the function will not be traced through and simply appear as a node in the graph.
-if is_torch_fx_available():
-    if not is_torch_greater_or_equal_than_1_13:
-        import torch.fx
+if is_flash_attn_2_available():
+    from flash_attn import flash_attn_func, flash_attn_varlen_func
+    from flash_attn.bert_padding import index_first_axis, pad_input, unpad_input  # noqa
+"""PyTorch Gemma model."""
```
Comment (ArthurZucker) [inline src/transformers/models/gemma/modeling_gemma.py:56]: this is potentially weird, I gotta remove this duplicate ruff is not able to

File: src/transformers/models/gemma/modeling_gemma.py [hunk 2]
```diff
@@ -55,25 +50,30 @@
...
 
 
 logger = logging.get_logger(__name__)
 
-_CONFIG_FOR_DOC = "GemmaConfig"
+
+ALL_LAYERNORM_LAYERS.append(GemmaRMSNorm)
+
+
+logger = logging.get_logger(__name__)
```
Comment (ArthurZucker) [inline src/transformers/models/gemma/modeling_gemma.py:69]: should be removed as well duplicate above

Comment (LysandreJik) [inline src/transformers/models/gemma/modeling_gemma.py:69]: :)

File: src/transformers/models/gemma/modeling_gemma.py (hunk 1 above)
Comment (LysandreJik) [reply src/transformers/models/gemma/modeling_gemma.py:56]: yes indeed, and sort imports back up as well (maybe ruff does that?)

File: src/transformers/models/gemma/modeling_gemma.py [hunk 3]
```diff
@@ -13,7 +13,6 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
```
Comment (LysandreJik) [inline src/transformers/models/gemma/modeling_gemma.py:15]: Please add something like:

```
# This file was automatically generated from <path_to_diff_file.py>.
# Do NOT edit this file manually as any edits will be overwritten by the generation of
# the file from the diff.
# If any change should be done, please apply the change to the diff.py file directly
```

or something along those lines

**Review by LysandreJik — APPROVED** [review]
Starting to look really really good!

**Review by amyeroberts — COMMENTED** [review]
Thanks for all the work on this - excited to see it in action! 

I've just done a first pass over the code. 

Out of interest, is there a reason that this approach was taken over the original `register` pattern that existed before? The construction of the diff files and the script I think are less clean. 

Two main comments: 
* custom methods should have proper typing and docstrings. I don't think most people will be familiar with libcst (I'm certainly not) and so it would help a lot to have a bit of guidance on the intention of the methods and what the objects are.
* Some of the logging / header makes assumptions about how people are viewing the files / reading output from the terminal. Let's avoid things which don't parse well in resized windows

File: src/transformers/models/gemma/diff_gemma.py [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+            inputs_embeds,
+            use_cache,
+            output_attentions,
+            output_hidden_states,
+            return_dict,
+            cache_position,
+        )
+
+
+# Example where we ony modify the docstring and call super
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/gemma/diff_gemma.py:415]: 💯 but not in this PR as it's kind of unrelated. We can quite simply change the `add_docstring` decorators to supports this!

File: utils/diff_model_converter.py [hunk 5]
```diff
@@ -0,0 +1,562 @@
...
+                expr,
+                m.SimpleStatementLine(
+                    body=[
+                        m.Return(
+                            value=m.Call(func=m.Attribute(value=m.Call(func=m.Name("super")), attr=m.Name(func_name)))
+                        )
+                    ]
+                ),
+            ):
+                new_body.extend(self.update_body(self.original_methods[func_name].body.body, node.body))
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply utils/diff_model_converter.py:307]: we can using `|` good catch

File: utils/diff_model_converter.py [hunk 6]
```diff
@@ -0,0 +1,562 @@
+import argparse
+import glob
+import importlib
+import re
+
+import libcst as cst
+from check_copies import fix_ruff
+from libcst import ClassDef, CSTTransformer, CSTVisitor
+from libcst import matchers as m
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply utils/diff_model_converter.py]: it is standard and a lot cleaner given the nested calls to it

File: utils/diff_model_converter.py [hunk 7]
```diff
@@ -0,0 +1,562 @@
...
+
+    def __init__(self, python_module):
+        self.python_module = python_module
+        self.classes = {}  # class LlamaAttentino
+        self.imports = {}  # from flash_attn import
+        self.function_def = {}  # def repeat_kv
+        self.assignments = {}  # LLAMA_DOCSTRING
+        self.class_dependency_mapping = {}  # "LlamaModel":["LlamaDecoderLayer, "LlamaRMSNorm", "LlamaPreTrainedModel"], "LlamaDecoderLayer":["LlamaAttention","Llama"]
+
+    def visit_ClassDef(self, node: ClassDef) -> None:
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply utils/diff_model_converter.py:91]: yep never used as is

File: utils/diff_model_converter.py [hunk 8]
```diff
@@ -0,0 +1,562 @@
...
+    ruffed_code = fix_ruff(new_mod.code)
+
+    with open(diff_file.replace("diff_", "modeling_"), "w") as f:
+        f.write(AUTO_GENERATED_MESSAGE + ruffed_code)
+
+    if hasattr(cst_transformers, "config_body"):
+        config_module = cst.Module(body=[*cst_transformers.config_body], header=new_mod.header)
+        with open(diff_file.replace("diff_", "configuration_"), "w") as f:
+            ruffed_code = fix_ruff(config_module.code)
+            f.write(AUTO_GENERATED_MESSAGE + ruffed_code)
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply utils/diff_model_converter.py]: Will do in another PR!

File: utils/diff_model_converter.py [hunk 9]
```diff
@@ -0,0 +1,562 @@
...
+        self.class_dependency_mapping = {}  # "LlamaModel":["LlamaDecoderLayer, "LlamaRMSNorm", "LlamaPreTrainedModel"], "LlamaDecoderLayer":["LlamaAttention","Llama"]
+
+    def visit_ClassDef(self, node: ClassDef) -> None:
+        self.classes[node.name.value] = node
+        for k in node.bases:  # deal with inheritance
+            name = self.python_module.code_for_node(k)
+            self.class_dependency_mapping.update(
+                {
+                    node.name.value: set(self.class_dependency_mapping.get(name, {name}))
+                    | self.class_dependency_mapping.get(node.name.value, set())
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply utils/diff_model_converter.py]: Updated to make it cleare

File: utils/diff_model_converter.py [hunk 10]
```diff
@@ -0,0 +1,562 @@
...
+
+    def visit_SimpleStatementLine(self, node):
+        match node:
+            case cst.SimpleStatementLine(body=[cst.Assign(targets=[_], value=_)]):
+                if m.matches(self.get_metadata(cst.metadata.ParentNodeProvider, node), m.Module()):
+                    self.assignments[node.body[0].targets[0].target.value] = node
+            case cst.SimpleStatementLine(body=[cst.Import(names=[_])]):
+                self.imports[node.body[0].names] = node
+            case cst.SimpleStatementLine(body=[cst.ImportFrom(_)]):
+                self.imports[node.body[0].names] = node
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply utils/diff_model_converter.py]: good catch

File: src/transformers/models/gemma/modeling_gemma.py [hunk 11]
```diff
@@ -55,25 +55,17 @@
...
 
 
-# This makes `_prepare_4d_causal_attention_mask` a leaf function in the FX graph.
-# It means that the function will not be traced through and simply appear as a node in the graph.
-if is_torch_fx_available():
-    if not is_torch_greater_or_equal_than_1_13:
-        import torch.fx
-
-    _prepare_4d_causal_attention_mask = torch.fx.wrap(_prepare_4d_causal_attention_mask)
+logger = logging.get_logger(__name__)
```
Comment (ArthurZucker) [inline src/transformers/models/gemma/modeling_gemma.py:58]: final diff!

**Comment by ArthurZucker** [comment]
@amyeroberts everything is adressed, merging and can adresse more stuff later on!

File: src/transformers/models/gemma/modeling_gemma.py [hunk 12]
```diff
@@ -397,7 +408,7 @@ def forward(
             query_states, key_states, value_states, attention_mask, q_len, dropout=dropout_rate
         )
 
-        attn_output = attn_output.reshape(bsz, q_len, -1).contiguous()
```
Comment (namin) [inline src/transformers/models/gemma/modeling_gemma.py:400]: I think this change is a regression. The shape doesn't seem to match.

Comment (ArthurZucker) [reply src/transformers/models/gemma/modeling_gemma.py:400]: yep updating this now!

**Comment by ArthurZucker** [comment]
FIxed by #31202

File: src/transformers/models/gemma/configuration_gemma.py [hunk 13]
```diff
@@ -1,5 +1,12 @@
+#           🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨
+#               This file was automatically generated from <path_to_diff_file.py>.
```
Comment (qubvel) [inline src/transformers/models/gemma/configuration_gemma.py:2]: Is it supposed to be a real path to a diff file or just `<path_to_diff_file.py>`?

Comment (ArthurZucker) [reply src/transformers/models/gemma/configuration_gemma.py:2]: just `<path_to_diff_file.py>` but we can also add the real path

**Pull request merged**